sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -213,20 +227,37 @@ data ProcessorMessage = Flush | Shutdown
 batchProcessor :: MonadIO m => BatchTimeoutConfig -> Exporter ImmutableSpan -> m Processor
 batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
   batch <- newIORef $ boundedMap maxQueueSize
-  workSignal <- newEmptyMVar
-  worker <- async $ loop $ do
-    req <-
-      liftIO $
-        timeout (millisToMicros scheduledDelayMillis) $
-          takeMVar workSignal
```

Comment:
Two possible problems with this code:

1. In an `uninterruptibleMask`, the `timeout` can never kill the `takeMVar`, and this could potentially never complete (until someone manually calls the flush or shutdown and the `MVar` is filled
2. It is possible for the `MVar` to be filled, and later calls to `tryPutMVar` fail. Suppose we had this flow:
    1. `MVar` is empty
    2. `takeMVar` blocks
    3. `tryPutMVar Flush` happens, `takeMVar` succeeds and proceeds to make HTTP calls
    4. `tryPutMVar Flush` happens, and succeeds
    5. `tryPutMVar Shutdown` happens, fails to write
    6. HTTP calls are done, `takeMVar` succeeds returning `Flush`.

Switching to two separate variables allows a separate "shutdown" and "flush now" flow. But using two variables means we want `STM`.